### PR TITLE
Enable virtio-iommu on AArch64 (FDT only)

### PR DESCRIPTION
**Pull request by michael2012z**
The PR enables virtio-iommu and its test case on AArch64.
Now virtio-iommu only works when FDT is used. The work on ACPI is on-going.

The PR includes:
- Adding IOMMU node in FDT
- Refactoring virtio-iommu and GIC code to receive architecture specific MSI-IOVA address
- Adapting the test case
- Documentation

### CODE REVIEW

**Review by MrXinWang — APPROVED** [review]

File: vmm/src/vm.rs [hunk 1]
```diff
@@ -1077,6 +1077,14 @@ impl Vm {
 
         let pci_space = (pci_space_start.0, pci_space_size);
 
+        let virtio_iommu_bdf = if let Some((virtio_iommu_bdf, _)) =
+            self.device_manager.lock().unwrap().iommu_attached_devices()
+        {
+            Some(*virtio_iommu_bdf)
+        } else {
+            None
+        };
```
Comment (sboeuf) [inline vmm/src/vm.rs:1086]: You can simplify with `map()` here:
```rust
let virtio_iommu_bdf = self.device_manager.lock().unwrap().iommu_attached_devices().map(|(v, _)| *v);
```

File: tests/integration.rs [hunk 2]
```diff
@@ -4017,28 +4016,47 @@ mod tests {
...
-                // Verify the first disk is located under IOMMU group 0.
+                // For AArch64, now virtual IOMMU is only tested on FDT, not ACPI.
+                // In the case of FDT, the behavior of IOMMU is a bit different with ACPI. All the
+                // devices on the PCI bus will be attached to the virtual IOMMU, except the
+                // virtio-iommu device itself. So these devices will all be added to IOMMU groups,
+                // and appear under folder '/sys/kernel/iommu_groups/'.
+                // The result is, on AArch64 IOMMU group '0' contains "0000:00:01.0" which is the
+                // console. The first disk "0000:00:02.0" is in group '1'.
+                // While on X86, console device is not attached to IOMMU. So the IOMMU group '0'
+                // contains "0000:00:02.0" which is the first disk.
```
Comment (sboeuf) [inline tests/integration.rs:4027]: I understand the difference of behavior for FDT, which sounds fine to me, but I'd like to see a note in the documentation `docs/iommu.fd` about it. Basically, for FDT, as soon as you enable the vIOMMU with one device, all the other devices are attached to it. This is not very intuitive wrt the CLI.
/cc @rbradford

File: vmm/src/vm.rs (hunk 1 above)
Comment (michael2012z) [reply vmm/src/vm.rs:1086]: Thank you for the graceful simplification.

File: tests/integration.rs (hunk 2 above)
Comment (michael2012z) [reply tests/integration.rs:4027]: Good idea. This special behavior should be mentioned in doc.

Comment (michael2012z) [reply tests/integration.rs:4027]: @sboeuf , I added a commit to update the document, PTAL.

**Review by sboeuf — APPROVED** [review]
Looks good, I'd like @rbradford's feedback about the special behavior of FDT before we merge it.

**Comment by rbradford** [comment]
> Looks good, I'd like @rbradford's feedback about the special behavior of FDT before we merge it.

Sounds fine!

**Pull request merged**